### DISCUSSION TITLE
Is creating a table via `Schema::CreateTableGenerator` supported behaviour?

### DISCUSSION THREAD

**Discussion by suetanvil** [jeremyevans/sequel#2205] [General]
I'm trying to create a table without using the DSL directly.  It looks like the way to do this is

1. Create a `Schema::CreateTableGenerator` and define the schema by calling its instance methods.
2. Invoke `Database::create_table` with the generator as its `:generator` option.

Looking at the source code, this looks like it will work (although I haven't tried it yet).  However, the documentation seems to indicate that `CreateTableGenerator` is internal and not something I should be messing with.  So my questions are:

1. Is the method I describe the preferred way of doing this or does Sequel provide another way I just don't know about?
2. Is this part of the public interface?  I.e. can I be confident that this will continue to work in the future?

For context, what I'm trying to do is create a class that behaves kind of like the Ruby `Struct` class but holds its data in a database.  I'd like users to be able to define the fields in the argument list and any methods in an optional block and avoid needing another block to evaluate the Sequel DSL.

(As an aside, I think Sequel is very cool in general and I really appreciate its existence.  I realize that what I'm doing is a bit outside of the problem domain.)

**Comment by jeremyevans** [comment]
If you look at the code history, you'll see not much has changed in the `Schema::CreateTableGenerator` in a long time, in terms of the API, so if you use it, it is unlikely things will break.  That being said, using it directly would be reaching into the internals a bit.  The preferred way would be to call `Database#create_table` with a block, but you already indicated you do not want to do that.  I can see where using the generator directly could be a little simpler.  However, storing the methods and arguments in an array and calling the methods inside the `create_table` block shouldn't be much more work.

**Reply by suetanvil** [reply]
Okay, thanks.  I'll probably go with your suggestion, just for the peace of mind.

As an aside, it occurs to me that it might be valuable to promote `CreateTableGenerator` to the public API. The use case would be for middleware that uses Sequel as a back end but wants to provide the DSL as part of its API.  It could then simply create a block, pass it to `CreateTableGenerator` and pass that to `create_table` while doing whatever else it needs to do.

Just a thought.

**Reply by jeremyevans** [reply]
I think few people would benefit from that, and it restricts long term flexibility.  Most people that could pass a block to `CreateTableGenerator` could pass it to `create_table`.  If you compare current Sequel to `3.0.0` (15 years ago), it shows that while the `create_table` API is the same, the generator integration is different (the class name changed: https://github.com/jeremyevans/sequel/blob/3.0.0/lib/sequel/database/schema_methods.rb#L61).
